measfunc/dmmqdac.py

```diff
@@ -2,182 +2,432 @@
+                           setpoints=(self.fast_channel_setpoints.voltage_setpoints,),
+                           parameter_class=Buffered1DAcquisition)
+
+        if (hasattr(self, 'slow_channel_setpoints')):
+            self.add_parameter('buffered_2d_acquisition',
+                            vals=Arrays(shape=(self.slow_channel_setpoints.num_samples, 
+                                                self.fast_channel_setpoints.num_samples)),
+                            setpoints=(self.slow_channel_setpoints.voltage_setpoints, 
+                                        self.fast_channel_setpoints.voltage_setpoints),
+                            parameter_class=Buffered2DAcquisition)
 
         self.add_parameter('sample_rate',
                            initial_value=1/0.003,
                            unit='Hz',
                            label='Sample Rate',
-                           vals=Numbers(1e-3,1),
+                           vals=Numbers(1,1e4),
                            get_cmd=None,
-                           set_cmd=None)
+                           set_cmd=None)  
+        
+        self.add_parameter('NPLC',
+                           initial_value=0.06,
+                           unit='a.u.',
+                           label='Number of power line cycles',
+                           vals=Numbers(),
+                           get_cmd=None,
+                           set_cmd=None) 
+
+        self.add_parameter('dmm_voltage_range',
+                           initial_value=1.0,
+                           unit='V',
+                           label='dmm voltage range',
+                           vals=Numbers(),
+                           get_cmd=None,
+                           set_cmd=None) 
 
-    def ramp_and_fetch(self):
-        self.setup_dmm()
+        self.add_parameter('num_repetitions',
+                           initial_value=1,
+                           unit='a.u.',
+                           label='Number of repetition averages',
+                           vals=Ints(),
+                           get_cmd=None,
+                           set_cmd=None) 
+
+    def ramp_voltages_2d_and_fetch_with_repetition_averaging(self):
+        data = np.zeros((self.slow_channel_setpoints.num_samples(), 
+                         self.fast_channel_setpoints.num_samples()))
+        for i_repetition in range(self.num_repetitions()):
+            data_i = self.ramp_voltages_2d_and_fetch_with_retry()
+            data += data_i/self.num_repetitions()
+        return data 
+
+    def ramp_voltages_2d_and_fetch_with_retry(self):
+        try:
+            data = self.ramp_voltages_2d_and_fetch()
+        except: # VisaIOError
+            self.setup_dmm_for_buffered_acquisition()
+            sleep(1)
+            data = self.ramp_voltages_2d_and_fetch()
+        return data 
+
+    def ramp_voltages_2d_and_fetch(self):
+        """
+        get data equivalent to 2d matrix (needs to be reshaped without doNds)
```

Comment:
I don't understand the last part of the comment "(needs to be reshaped without doNds)"